src/bigistore.c

```diff
@@ -647,6 +648,185 @@ Datum bigistore_each(PG_FUNCTION_ARGS)
+    BigIStorePair   **pairs;
+    ReturnSetInfo   *rsinfo;
+
+    prepTuplestoreResult(fcinfo);
+
+    /* get a tuple descriptor for our result type */
+    switch (get_call_result_type(fcinfo, NULL, &tupdesc))
+    {
+        case TYPEFUNC_COMPOSITE:
+            /* success */
+            break;
+        case TYPEFUNC_RECORD:
+            /* failed to determine actual type of RECORD */
+            ereport(ERROR,
+                    (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+                errmsg("function returning record called in context "
+                       "that cannot accept type record")));
+            break;
+        default:
+            /* result type isn't composite */
+            elog(ERROR, "return type must be a row type");
+            break;
+    }
+
+    rsinfo = (ReturnSetInfo *) fcinfo->resultinfo;
+    Assert(rsinfo->returnMode == SFRM_Materialize);
+
+    oldcontext          = MemoryContextSwitchTo(rsinfo->econtext->ecxt_per_query_memory);
+    tupstore            = tuplestore_begin_heap(true, false, work_mem);
+    rsinfo->setResult   = tupstore;
+    rsinfo->setDesc     = tupdesc;
+    MemoryContextSwitchTo(oldcontext);
+
+    /* deconstruct argument array */
+    input       = PG_GETARG_ARRAYTYPE_P(0);
+    i_eltype    = ARR_ELEMTYPE(input);
+    get_typlenbyvalalign(i_eltype, &i_typlen, &i_typbyval, &i_typalign);
+    deconstruct_array(input, i_eltype, i_typlen, i_typbyval, i_typalign,
+                      &i_data, &i_nulls, &arg_count );
+
+    offset  = palloc0(sizeof(int) * arg_count);
+    len     = palloc0(sizeof(int) * arg_count);
+    pairs   = palloc0(sizeof(BigIStorePair *) * arg_count);
+    dvalues = palloc0(sizeof(Datum) * (arg_count + 1));
+    nulls   = palloc0(sizeof(bool) * (arg_count + 1));
+
+    for(int i = 0; i < arg_count; i++ )
+    {
+        BigIStore *bigistore = (BigIStore *) i_data[i];
+
+        len[i]      = bigistore->len;
+        pairs[i]    = FIRST_PAIR(bigistore, BigIStorePair);
+    }
+
+    for(;;)
+    {
+        int32 current_key;
+
+        memset(nulls, false, (arg_count+1) * sizeof(bool));
+        /* set first key to null to indicate end */
+        nulls[0] = true;
+
+        for(int i = 0; i < arg_count; i++ )
+        {
+            if(offset[i] >= len[i])
+            {
+                nulls[i+1] = true;
+                continue;
+            }
+            else
+            {
+                if(nulls[0])
+                {
+                    nulls[0] = false;
+                    current_key = pairs[i]->key;
+                }
+            }
+
+            /* when key matches take it */
+            if(current_key == pairs[i]->key)
+            {
+                dvalues[i+1] = Int64GetDatum(pairs[i]->val);
+                offset[i]++;
+                pairs[i]++;
+            }
+            /* if current_key is to big, all already set non null values must be reset */
+            else if(current_key > pairs[i]->key)
```

Comment:
why do we compare for greater than, but not less than (the equal case is covered above). I assume this code produces the join result, but it seems complicated. Perhaps a comment would help...

Reply:
we have equal, greater than and else so imho all cases are covered

Reply:
We do, but I couldn't figure out why do we specifically compare for greater then and leave the else for the less then. A comment before the outer for loop would definitely help.

Reply:
true let me think of a good comment